WIP Separate platform & target

sdmaclea opened the pull request:
Extend cross component build make system to allow cross OS build for some components.

Standardize on `Platform` and `Target` for the nomenclature (rename `_Host_` to `_Platform_`)
Add `CLR_CMAKE_TARGET_*`
Modify feature flag definition to depend on `CLR_CMAKE_TARGET_*`
Modify crosscomponent make to handle `CLR_CMAKE_TARGET_OS` 
Set `CLR_CMAKE_TARGET_OS` in build.cmd
Add flag to set target OS to linux.
Remove stray sos reference

sdmaclea committed:
Define CLR_CMAKE_TARGET_OS for __BuildCrossArchNative

sdmaclea committed:
Remove stray sos reference

sdmaclea committed:
Add option -linux to build.cmd

sdmaclea committed:
Enable Linux x64 cross in build.cmd

sdmaclea committed:
Add CLR_CMAKE_PLATFORM_OS

sdmaclea committed:
Rename CLR_CMAKE_HOST_ARCH CLR_CMAKE_PLATFORM_ARCH

sdmaclea committed:
Set CLR_CMAKE_TARGET_UNIX variables

sdmaclea committed:
Add CLR_CMAKE_PLATFORM_OS to the cross message

sdmaclea committed:
Make clrdefinitions.cmake depend on TARGET OS properties

sdmaclea committed:
Add PLATFORM_ and TARGET_ definitions

sdmaclea committed:
Remove unnecessary defines ARM, ARM64, and AMD64

sdmaclea committed:
Add TARGET_WINDOWS

janvorli commented:
@sdmaclea it is not clear to me how we can ever build native components cross os, can you please share more details on how are we planning to do that?

sdmaclea commented:
@janvorli This work is experimental at this point.

sdmaclea commented on src/coreclr/build.cmd:
```diff
@@ -128,6 +128,8 @@ if /i "%1" == "-x86"                 (set __BuildArchX86=1&set processedArgs=!pr
 if /i "%1" == "-arm"                 (set __BuildArchArm=1&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 if /i "%1" == "-arm64"               (set __BuildArchArm64=1&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
 
+if /i "%1" == "-linux"               (set __BuildOS=Linux&set processedArgs=!processedArgs! %1&shift&goto Arg_Loop)
```
This option may imply general support.  Maybe a special purpose flag would be better.

sdmaclea committed:
Remove -linux add -linuxdac

sdmaclea committed:
FEATURE_PAL

Make FEATURE_PAL depend on the platform

sdmaclea committed:
Set __CMakeNeeded for -linuxdac

sdmaclea committed:
Disable mscordbi on cross OD builds

sdmaclea committed:
Add FEATURE_REMOTE_PROC_MEM

sdmaclea committed:
ShimRemoteDataTarget::GetPlatform use TARGET_UNIX

sdmaclea committed:
Fix conversion warning as error

sdmaclea committed:
Fix warning as error

sdmaclea committed:
Hack: Disable failing assert

sdmaclea committed:
TARGET_UNIX daccess.cpp

sdmaclea committed:
Hack: DAC_TABLE_RVA

sdmaclea committed:
TARGET_UNIX GetPlatform

sdmaclea committed:
TARGET_UNIX MAKEDLLNAME

sdmaclea committed:
Move ThreadStore::m_Crst

Move the Crst member to the end of the structure for the cross OS DAC.

sdmaclea committed:
Adjust DAC_TABLE_RVA

sdmaclea committed:
Separate WIN32 from TARGET_WIN32

sdmaclea committed:
TARGET_UNIX dacvars.h

sdmaclea committed:
TARGET_UNIX vptr_list.h

sdmaclea committed:
TARGET_UNIX request.cpp

sdmaclea committed:
TARGET_UNIX DISABLE_CONTRACTS

sdmaclea committed:
THREAD_UNIX Thread

sdmaclea committed:
Hack Crst size

sdmaclea committed:
SHash correct layout

Remove multiple inheritance by dropping noncopyable
This corrects the layout of the SHash structure on Windows to
match Linux.

The multiple inheritance was in serting an extra byte of padding in
the structure layout.

sdmaclea commented:
This prototype worked, but this code needs to be polished before merging.  Closing this in favor of smaller actionable PRs.
